src/main/java/uk/gov/justice/hmiprobation/casesampler/utils/RoundingAdjuster.kt

```diff
@@ -0,0 +1,51 @@
+package uk.gov.justice.hmiprobation.casesampler.utils
+
+import uk.gov.justice.hmiprobation.casesampler.utils.Type.DECREASED_FOR_ROUNDING
+import uk.gov.justice.hmiprobation.casesampler.utils.Type.INCREASED_FOR_ROUNDING
+
+fun <K> adjustForRounding(numberOfSamples: Int, results: SampleSizes<K>): SampleSizes<K> {
+    val totalSampleSize = totalSize(results)
+    return when {
+        results.isEmpty() -> results
+        totalSampleSize < numberOfSamples -> addAdditionalSamples(numberOfSamples - totalSampleSize, results)
+        totalSampleSize > numberOfSamples -> removeAdditionalSamples(totalSampleSize - numberOfSamples, results)
+        else -> results
+    }
+}
+
+private fun <K> totalSize(results: SampleSizes<K>): Int = results.sumBy { (_, size) -> size.count }
+
+private fun <K> removeAdditionalSamples(samplesToRemove: Int, results: SampleSizes<K>): SampleSizes<K> {
+    val keyToSize = results.toMap().toMutableMap()
+    val infiniteKeys = infiniteKeysIterator(results.sortedByDescending { (_, size) -> size.count })
```

Comment:
Just wondering why the name infiniteKeys? Are your `results` not finite?

Reply:
the results are but this creates a sequence that is seeded by the passed in collection once exhauster. This would be similar to using a [cycle iterator in guava](https://guava.dev/releases/15.0/api/docs/com/google/common/collect/Iterators.html#cycle(T...))
 
so given 3 groups (ordered from largest to smallest) and 5 samples to remove we would visit the groups in order:
1, 2, 3, 1, 2